Hoodie Event callbacks

kaushikd49 opened the pull request:
Addresses #153 (Implement Hoodie Event callbacks)

kaushikd49 commented:
CI failure is due to the race condition bug on master. This PR has been successfully tested on master with the patch given by @n3nash to avoid race temporarily. PR has has also been tested with an ingestion job.

prazanna commented on hoodie-client/src/main/java/com/uber/hoodie/config/HoodieWriteConfig.java:
```diff
@@ -102,6 +105,17 @@ public StorageLevel getWriteStatusStorageLevel() {
 
+    public <T extends WriteStatus> T getWriteStatusInstance() {
+        try {
+            Class<? extends WriteStatus> aClass = (Class<? extends WriteStatus>) Class
```
Use ReflectionUtils to cache the class instance and config is not the right place to instantiate the class. I would do it as a interface method (factory) in WriteStatus interface

kaushikd49 replied:
Sure Prasanna, will do.

vinothchandar replied:
I agree with what prasanna is saying. We can't be actually instantiating the class from the config class.  i.e the method here can just return you the class name.. 

Please move the actual creation of the class i.e the ReflectionUtils call into the Handle classes themselves.

prazanna commented on hoodie-client/src/main/java/com/uber/hoodie/WriteStatus.java:
```diff
@@ -47,12 +48,14 @@
 
-    public void markSuccess(HoodieRecord record) {
+    public void markSuccess(HoodieRecord record,
+        Map<String, String> recordMetadata) {
```
Make WriteStatus a interface and have a Default implementation

kaushikd49 replied:
Sure, Im planning to have the default implementation also not use recordMetadata. Please let me know if you prefer otherwise.

prazanna commented on hoodie-common/src/main/java/com/uber/hoodie/common/model/HoodieRecordPayload.java:
```diff
@@ -56,4 +57,10 @@
+     * This method can be used to fetch metadata regarding the HoodieRecordPayload, for the purpose
+     * of capturing some metrics.
+     */
+    Map<String, String> getMetadata();
```
I am a little concerned with having a map of properties on the record payload. Adding data to record payload is very tricky in terms of shuffle. It put in a 100 properties and tuned it for 10 - things may fail miserably during a shuffle. If we need to encode data - why dont we just do it in the InsertRecord. Can you may be give me an example where storing it in the record itself would not make sense?

kaushikd49 replied:
CMIIW, even if we encode it in the InsertRecord (or HoodieRecordPayload), we need this method to fetch the metadata just before HoodieRecordPayload is deflated like in https://github.com/uber/hoodie/pull/251/commits/8f2c88b836b25d5bd66859abcc2dd49d07bb3e28#diff-165908ecbfe7dbae538c00e8157e7be0R129

Regarding storing properties in a map - did you recommend storing it in the IndexedRecord instead? My concern was inadvertent stuff being put into IndexRecord affecting parquet schema. But we are planning to do that anyway for across_batch_de-duped counts, so yes, I could actually move these properties into IndexedRecord itself.

vinothchandar replied:
IIUC you want a mechanism to pass some data from each record and have it propagated down to a custom WriteStatus class? Since it was mostly you and @prazanna who discussed this and neither the ticket nor the PR describes what the commit is about, @kaushikd49 could you please summarize what the real use-case is ?

prazanna replied:
I am okay with this approach @kaushikd49. We may want to enable adding properties to HoodieRecordPayload with a property in HoodieWriteConfig (disabled by default). It may be good idea to write up a small doc on how you are planning to use this map downstream internally and share it with @vinothchandar - just so he has context. Thanks and sorry for the delay

kaushikd49 replied:
@vinothchandar Yea that is right. I will be sharing a doc with you shortly explaining the use-case.

@prazanna Sure, I'll do that.

kaushikd49 commented:
@prazanna Thanks a lot for your comments. Had few questions on the comments. Could you please help answer them?

vinothchandar commented:
@kaushikd49 I will take a look and in the meantime, can you please rebase this again off master

kaushikd49 commented:
Thanks, done @vinothchandar

vinothchandar commented on hoodie-client/src/test/java/com/uber/hoodie/common/TestRawTripPayload.java:
```diff
@@ -91,6 +97,14 @@ public String getPartitionPath() {
 
+    @Override
+    public Map<String, String> getMetadata() {
+        Map<String, String> metadataMap = new HashMap<>();
```
Can you come up with some reasonable values here for the map? Current values don't convey your usage pattern well IMO

vinothchandar commented on hoodie-client/src/test/java/com/uber/hoodie/table/TestCopyOnWriteTable.java:
```diff
@@ -104,7 +105,7 @@ private HoodieWriteConfig makeHoodieClientConfig() throws Exception {
         // Prepare the AvroParquetIO
         String schemaStr = IOUtils.toString(getClass().getResourceAsStream("/exampleSchema.txt"), "UTF-8");
-        return HoodieWriteConfig.newBuilder().withPath(basePath).withSchema(schemaStr);
+        return HoodieWriteConfig.newBuilder().withPath(basePath).withSchema(schemaStr).withWriteStatusClass(SampleMergeMetadataWriteStatus.class);
```
Please write a separate test case, which demonstrates how the framework should be used. Lets not change the existing tests with a custom WriteStatus class, as I want to ensure the default usage with WriteStatus.class continues to be exercised.

vinothchandar commented on hoodie-common/src/main/java/com/uber/hoodie/common/HoodieJsonPayload.java:
```diff
@@ -57,6 +59,11 @@ public HoodieJsonPayload(String json) throws IOException {
         return Optional.of(jsonConverter.convert(getJsonData()));
     }
 
+    @Override
```
Can you push this to the base class, instead of implementing in each subclass

vinothchandar commented on hoodie-common/src/main/java/com/uber/hoodie/common/model/HoodieRecordPayload.java:
```diff
@@ -56,4 +57,10 @@
+     * This method can be used to fetch metadata regarding the HoodieRecordPayload, for the purpose
+     * of capturing some metrics.
+     */
+    Map<String, String> getMetadata();
```
In fact, can you make this an `Optional<Map<String, String>>` and always return Optional.empty from the base class?

vinothchandar commented on hoodie-client/src/main/java/com/uber/hoodie/WriteStatus.java:
```diff
@@ -47,12 +48,14 @@
     private long totalErrorRecords = 0;
 
-    public void markSuccess(HoodieRecord record) {
+    public void markSuccess(HoodieRecord record,
```
javadocs for parameters..

kaushikd49 commented:
@vinothchandar @prazanna Thank you, will send out an updated PR

kaushikd49 commented:
@vinothchandar @prazanna Addressed your comments. 

> We may want to enable adding properties to HoodieRecordPayload with a property in HoodieWriteConfig (disabled by default)

Did not add this property because in the default WriteStatus implementation class (WriteStatusImpl), the metadata map is anyway not used. Also, this adds a dependency to HoodieWriteConfig class from WriteStatus Interface. Open to your suggestions on this.

vinothchandar commented on hoodie-client/src/main/java/com/uber/hoodie/io/HoodieAppendHandle.java:
```diff
@@ -74,7 +73,7 @@ public HoodieAppendHandle(HoodieWriteConfig config,
                               Iterator<HoodieRecord<T>> recordItr) {
         super(config, commitTime, hoodieTable);
-        WriteStatus writeStatus = new WriteStatus();
+        WriteStatus writeStatus = ReflectionUtils.loadClass(config.getWriteStatusClassName());
```
If you re going to rely on a constructor being called via Reflection, then making `WriteStatus` an interface is not a necessary thing.. Atleast in the subclassing route, you will be forcing the subclass to implement atleast 1 constructor of the same signature as base class

Was this part of @prazanna 's suggestion earlier on?  (given then comment history is nt there now)

kaushikd49 replied:
@vinothchandar Yea, it was part of @prazanna's suggestion earlier. The rebase has hidden the comment. Ok, I'll retain WriteStatus like before as a class.

vinothchandar commented on hoodie-client/src/test/java/com/uber/hoodie/TestMergeOnReadTable.java:
```diff
@@ -224,7 +225,29 @@ public void testSimpleInsertAndUpdate() throws Exception {
+        HoodieWriteConfig cfg = getConfigBuilder().withWriteStatusClass(MetadataMergeWriteStatus.class).build();
+        HoodieWriteClient client = new HoodieWriteClient(jsc, cfg);
+
+        /**
```
there is only 1 write.. please remove comment block.

vinothchandar commented:
Generally okay with the change. 1 question on interfacing vs subclassing. Also please ensure you rebased against latest master, since I changed WriteStatus yest

kaushikd49 commented:
@vinothchandar Done. Yea it is updated with master.

vinothchandar commented on hoodie-client/src/main/java/com/uber/hoodie/WriteStatus.java:
```diff
@@ -24,112 +24,132 @@
 public class WriteStatus implements Serializable {
 
-    private final HashMap<HoodieKey, Throwable> errors = new HashMap<>();
+  private final HashMap<HoodieKey, Throwable> errors = new HashMap<>();
```
Indenting has changed here throughout.. Can you reformat this file and ensure the only diffs we see here are the two methods?  https://uber.github.io/hoodie/code_structure.html

kaushikd49 replied:
Sorry about that. 

Im using the same code style, and it seems like it is using +2 spaces for indentation.  https://google.github.io/styleguide/javaguide.html#s4.2-block-indentation

vinothchandar commented on hoodie-client/src/test/java/com/uber/hoodie/table/TestCopyOnWriteTable.java:
```diff
@@ -238,7 +239,40 @@ public void testUpdateRecords() throws Exception {
     }
 
-    @Test public void testInsertWithPartialFailures() throws Exception {
+    // Check if record level metadata is aggregated properly at the end of write.
```
watch for indentation here too

vinothchandar commented on hoodie-common/src/main/java/com/uber/hoodie/common/HoodieJsonPayload.java:
```diff
@@ -20,6 +20,8 @@
 import com.uber.hoodie.common.model.HoodieRecordPayload;
 import com.uber.hoodie.exception.HoodieException;
 
+import java.util.Collections;
```
extra changes that are not needed?

kaushikd49 commented:
@vinothchandar Done.
